File: pamo_c/src/stage3/spd_project.c
/*
 * Copyright 2024 Light Transport Entertainment Inc.
 * SPDX-License-Identifier: Apache-2.0
 *
 * 9x9 SPD (symmetric positive semi-definite) matrix projection
 * via Jacobi eigenvalue iteration.
 *
 * Algorithm: eigendecompose, clamp negative eigenvalues to 0, reconstruct.
 * Based on Burkardt's Jacobi eigenvalue implementation (GNU LGPL).
 */
#include "pamo/pamo_types.h"

#include <math.h>
#include <string.h>

void pamo_spd_project_9x9(double a[9][9], int max_iters) {
    double d[9], v[9][9], bw[9], zw[9];

    /* Initialize V = identity. */
    memset(v, 0, sizeof(v));
    for (int i = 0; i < 9; i++) v[i][i] = 1.0;

    /* D = diagonal of A. */
    for (int i = 0; i < 9; i++) {
        d[i] = a[i][i];
        bw[i] = d[i];
        zw[i] = 0.0;
    }

    for (int it = 0; it < max_iters; it++) {
        /* Convergence check: upper triangle norm. */
        double thresh = 0.0;
        for (int j = 0; j < 9; j++)
            for (int i = 0; i < j; i++)
                thresh += a[i][j] * a[i][j];
        thresh = sqrt(thresh) / 36.0;
        if (thresh == 0.0) break;

        for (int p = 0; p < 9; p++) {
            for (int q = p + 1; q < 9; q++) {
                double gapq = 10.0 * fabs(a[p][q]);
                double termp = gapq + fabs(d[p]);
                double termq = gapq + fabs(d[q]);

                if (it > 3 && termp == fabs(d[p]) && termq == fabs(d[q])) {
                    a[p][q] = 0.0;
                } else if (thresh <= fabs(a[p][q])) {
                    double h = d[q] - d[p];
                    double term = fabs(h) + gapq;
                    double t;
                    if (term == fabs(h)) {
                        t = a[p][q] / h;
                    } else {
                        double theta = 0.5 * h / a[p][q];
                        t = 1.0 / (fabs(theta) + sqrt(1.0 + theta * theta));
                        if (theta < 0.0) t = -t;
                    }

                    double c = 1.0 / sqrt(1.0 + t * t);
                    double s = t * c;
                    double tau = s / (1.0 + c);
                    h = t * a[p][q];

                    zw[p] -= h;
                    zw[q] += h;
                    d[p] -= h;
                    d[q] += h;
                    a[p][q] = 0.0;

                    for (int j = 0; j < p; j++) {
                        double g = a[j][p], hh = a[j][q];
                        a[j][p] = g - s * (hh + g * tau);
                        a[j][q] = hh + s * (g - hh * tau);
                    }
                    for (int j = p + 1; j < q; j++) {
                        double g = a[p][j], hh = a[j][q];
                        a[p][j] = g - s * (hh + g * tau);
                        a[j][q] = hh + s * (g - hh * tau);
                    }
                    for (int j = q + 1; j < 9; j++) {
                        double g = a[p][j], hh = a[q][j];
                        a[p][j] = g - s * (hh + g * tau);
                        a[q][j] = hh + s * (g - hh * tau);
                    }
                    for (int j = 0; j < 9; j++) {
                        double g = v[j][p], hh = v[j][q];
                        v[j][p] = g - s * (hh + g * tau);
                        v[j][q] = hh + s * (g - hh * tau);
                    }
                }
            }
        }

        for (int i = 0; i < 9; i++) {
            bw[i] += zw[i];
            d[i] = bw[i];
            zw[i] = 0.0;
        }
    }

    /* Reconstruct A = V * max(D, 0) * V^T */
    memset(a, 0, 9 * 9 * sizeof(double));
    for (int k = 0; k < 9; k++) {
        if (d[k] <= 0.0) continue;
        for (int i = 0; i < 9; i++)
            for (int j = 0; j < 9; j++)
                a[i][j] += d[k] * v[i][k] * v[j][k];
    }

    /* Enforce symmetry. */
    for (int i = 0; i < 9; i++)
        for (int j = i + 1; j < 9; j++) {
            double avg = (a[i][j] + a[j][i]) * 0.5;
            a[i][j] = avg;
            a[j][i] = avg;
        }
}

File: simp_cuda/safe_project/src/pamo_safe_project/kernels/spd_project.py
"""
Pure warp implementation of spd_project_blocks, replacing the custom
built-in from the Rabbit-Hu/warp fork (warp 1.0.0-beta.2).

Algorithm: Jacobi eigenvalue iteration on a 9x9 symmetric matrix,
clamp negative eigenvalues to zero, reconstruct.
Based on John Burkardt's implementation (GNU LGPL).
"""
import warp as wp

mat99 = wp.types.matrix(shape=(9, 9), dtype=float)
vec9 = wp.types.vector(length=9, dtype=float)


@wp.func
def _jacobi_eigenvalue_9x9(a: mat99, it_max: int):
    n = 9

    v = mat99()
    for i in range(9):
        v[i, i] = 1.0

    d = vec9()
    bw = vec9()
    zw = vec9()
    for i in range(9):
        d[i] = a[i, i]
        bw[i] = d[i]
        zw[i] = 0.0

    for it in range(it_max):
        thresh = float(0.0)
        for j in range(9):
            for i in range(j):
                thresh = thresh + a[i, j] * a[i, j]
        thresh = wp.sqrt(thresh) / float(4 * n)

        if thresh == 0.0:
            break

        for p in range(9):
            for q in range(p + 1, 9):
                gapq = 10.0 * wp.abs(a[p, q])
                termp = gapq + wp.abs(d[p])
                termq = gapq + wp.abs(d[q])

                if it > 3 and termp == wp.abs(d[p]) and termq == wp.abs(d[q]):
                    a[p, q] = 0.0
                elif thresh <= wp.abs(a[p, q]):
                    h = d[q] - d[p]
                    term = wp.abs(h) + gapq

                    if term == wp.abs(h):
                        t = a[p, q] / h
                    else:
                        theta = 0.5 * h / a[p, q]
                        t = 1.0 / (wp.abs(theta) + wp.sqrt(1.0 + theta * theta))
                        if theta < 0.0:
                            t = -t

                    c = 1.0 / wp.sqrt(1.0 + t * t)
                    s = t * c
                    tau = s / (1.0 + c)
                    h = t * a[p, q]

                    zw[p] = zw[p] - h
                    zw[q] = zw[q] + h
                    d[p] = d[p] - h
                    d[q] = d[q] + h

                    a[p, q] = 0.0

                    for j in range(p):
                        g = a[j, p]
                        hh = a[j, q]
                        a[j, p] = g - s * (hh + g * tau)
                        a[j, q] = hh + s * (g - hh * tau)

                    for j in range(p + 1, q):
                        g = a[p, j]
                        hh = a[j, q]
                        a[p, j] = g - s * (hh + g * tau)
                        a[j, q] = hh + s * (g - hh * tau)

                    for j in range(q + 1, 9):
                        g = a[p, j]
                        hh = a[q, j]
                        a[p, j] = g - s * (hh + g * tau)
                        a[q, j] = hh + s * (g - hh * tau)

                    for j in range(9):
                        g = v[j, p]
                        hh = v[j, q]
                        v[j, p] = g - s * (hh + g * tau)
                        v[j, q] = hh + s * (g - hh * tau)

        for i in range(9):
            bw[i] = bw[i] + zw[i]
            d[i] = bw[i]
            zw[i] = 0.0

    return d, v


@wp.func
def spd_project_blocks(blocks: wp.array(dtype=wp.mat33, ndim=2), it_max: int):
    """
    Project the top-left 3x3 sub-grid of mat33 blocks (= 9x9 matrix)
    to positive semi-definite. Operates in-place on the 2D slice.
    Drop-in replacement for wp.spd_project_blocks(3, blocks, it_max).
    """
    # blocks_to_array
    a = mat99()
    for bi in range(3):
        for bj in range(3):
            blk = blocks[bi, bj]
            for k in range(3):
                for l in range(3):
                    a[bi * 3 + k, bj * 3 + l] = blk[k, l]

    d, v = _jacobi_eigenvalue_9x9(a, it_max)

    # Reconstruct with positive eigenvalues only
    result = mat99()
    for kk in range(9):
        if d[kk] > 0.0:
            for i in range(9):
                for j in range(9):
                    result[i, j] = result[i, j] + d[kk] * v[kk, i] * v[kk, j]

    # array_to_blocks with symmetry enforcement
    for bi in range(3):
        for bj in range(3):
            blk = wp.mat33()
            for k in range(3):
                for l in range(3):
                    ia = bi * 3 + k
                    ja = bj * 3 + l
                    blk[k, l] = (result[ia, ja] + result[ja, ia]) * 0.5  # type: ignore
            blocks[bi, bj] = blk
